// hotLogger/lib/hotLogger.js
/*!
 * Module dependencies.
 */

var hotplate = require('hotplate')
;

exports.log = function(entry){

  // Allow modules to manipulate the blog entry
  hotplate.invokeAll('aboutToLog', entry);

  // Allow modules to actually log this. Note that this module
  // doesn't actually do any logging itself
  hotplate.invokeAll('log', entry);
}


exports.hotHooks = hooks = {};

hooks.aboutToLog = function(entry){

  // Set some defaults. This hook is called by the module itself, as
  // this is the default way of adding fields!
  entry.logLevel   = entry.logLevel  ? entry.logLevel  : 0;
  entry.errorName  = entry.errorName ? entry.errorName : '';
  entry.message    = entry.message   ? entry.message   : '';
  entry.data       = entry.data      ? entry.data      : {};
  entry.loggedOn = new Date();
}

hooks.logFields = function(){
  return {
    logLevel:  { type: Number },
    errorName: { type: String, index: true  },
    message:   { type: String, index: true  },
    data:      { type: String, index: false, serialize: true },
    loggedOn:  { type: Date,   index: true  },
  };
}
