// hotServerLogger/lib/hotServerLogger.js
/*!
 * Module dependencies.
 */

var dummy
, hotplate = require('hotplate')
, path = require('path')
;

exports.log = function(entry){

  // Allow modules to manipulate the log entry, calling "about to blog" and then "blog"
  hotplate.invokeAll('aboutToLog', entry , function( ){
    hotplate.invokeAll('log', entry, null );
  });

}


exports.hotHooks = hooks = {};

hooks.aboutToLog = function( done, entry){

  // Set some defaults. This hook is called by the module itself, as
  // this is the default way of adding fields!
  entry.logLevel   = entry.logLevel  ? entry.logLevel  : 0;
  entry.errorName  = entry.errorName ? entry.errorName : '';
  entry.message    = entry.message   ? entry.message   : '';
  entry.data       = entry.data      ? entry.data      : {};
  entry.loggedOn = new Date();

  done( null );

}

hooks.logFields = function( done ){
  done( null, {
    logLevel:  { type: Number },
    errorName: { type: String, index: true  },
    message:   { type: String, index: false  },
    data:      { type: String, index: false, serialize: true },
    loggedOn:  { type: Date,   index: true  },
  } );
}

// ...or like this (elements in page)
hooks.pageElements = function( done ){
  done( null, { moduleName: 'hotServerLogger', result:{
    jses:  [ 'logger.js' ],
  } } );
}

// ...or like this (client paths)
hooks.clientPaths = function( done ){
  done( null, { moduleName: 'hotServerLogger', result: [ path.join(__dirname, '../client') ] } );
}
