// hotMongoLogger/lib/hotMongoLogger.js

var util = require('util')

  , hotplate = require('hotplate')
  , app = hotplate.app
  , db = hotplate.get('db')

  , ObjectId = require('mongowrapper').ObjectId

  , declare = require('simpledeclare')

  , Store = require('jsonreststores-mongo')
  , Schema = Store.Schema
;


// Module's variables
// var logFields = {};
var LogStore = null;

exports.hotHooks = hooks = {}

hooks.initLoggers = function( done, logFields ){

  var s = new Schema( logFields );

  // MAYBE: Manipulate schema here if needed: add things so that they are marked as sortable etc.

  LogStore = declare( Store,  {
    storeName: 'logs',
    db: db,
    schema: s,
    paramIds: [ 'none' ],
    handleGetQuery: true,
  });

  // TODO: Uncomment this when simpleschema-mongo is completed
  // s.buldIndexes(); 

  done( null ); 
}

hooks.run = function( done ){

  Store.make.All( app, '/call/:workspaceIdCall/logs', ':logId', LogStore );
  done( null );
}

hooks.log = function( done, entry){

  console.log("hotMongoLogger is logging:");
  console.log( entry );

  // Simply insert
  var logs = db.collection('logs').insert( entry, function() {}  );

  done( null );

}

hooks.stores = function( done ){
  done( null, {
    logs: { target: '/call/:workspaceIdCall/logs/', idProperty: '_id', sortParam: 'sortBy', },
  } );
}
