// hotMessages/lib/hotMessages.js
var 
  dummy
, hotplate =  require('hotplate')
, path = require('path')
, mongoose = require('mongoose')
, Schema = mongoose.Schema
, ObjectId = Schema.Types.ObjectId
, async = require('async')
, util = require('util')
;

app = hotplate.app;


exports.hotHooks = hooks = {}

hooks.init = function( done ){


  // Sets the Workspace schema, which is simply a list with name, active flag,
  // and the list of users who can access it (though the [Access] schema)
  var Tab = new Schema({
    // _id            : { type: ObjectId, index: true },  // TODO: Why if this is defined, it doesn't get created?
    userId         : { type: ObjectId, index: true }, 
    workspaceId    : { type: ObjectId, index: true }, 
    lastPing       : { type: Date, index: true },

    messages: [{
      from      : { type: ObjectId, index: true },
      type      : { type: String,   index: true },
      message   : { type: Schema.Types.Mixed },
    }],
  });

  mongoose.model('Tab', Tab );

  done( null );
}


hooks.run = function( done ){
  app.post( '/call/:workspaceIdCall/ping/:tabId', pollPing );
  app.get( '/call/:workspaceIdCall/register', registerPing );

  done( null );
}

// Simply activate path to client files
hooks.clientPaths = function( done ){
  done( null, { moduleName: 'hotMessages', result: [ path.join(__dirname, '../client') ] } );
}


function registerTab(userId, workspaceId, cb){
  var Tab = mongoose.model('Tab');

  var tab = new Tab();
  tab.userId = userId;
  tab.workspaceId = workspaceId;
  tab.lastPing = new Date();
  tab.messages = [];

  tab.save( cb );
}

function unregisterOldTabs( cb ){
  var Tab = mongoose.model('Tab');

  var fiveMinutesAgo = new Date();
  fiveMinutesAgo.setMinutes( fiveMinutesAgo.getMinutes() - 5 );

  Tab.remove( { lastPing: { $lt: fiveMinutesAgo } }, cb );
}

/*
function unregisterTab( tabId, cb ){
  var Tab = mongoose.model('Tab');

  Tabs.remove( { _id: tabId }, cb );
}
*/

function retrieveTab( tabId, cb ){
  var Tab = mongoose.model('Tab');

  Tab.findOneAndUpdate( { _id: tabId }, { lastPing: new Date() }, cb );
}

function sendToTabsOfWorkspace(toWorkspaceId, excludeTabId, fromUserId, type, message, cb){
  var Tab = mongoose.model('Tab');

  // The callback needs to be there full stop
  cb = cb || function(){} 

  var fiveMinutesAgo = new Date();
  fiveMinutesAgo.setMinutes( fiveMinutesAgo.getMinutes() - 5 );

  // Tab.update( { workspaceId: toWorkspaceId, tabId: { $ne: excludeTabId}, lastPing: { $gt: new Date() - 60*5*1000  } },
  Tab.update( { workspaceId: toWorkspaceId, _id: { $ne: excludeTabId}, lastPing: { $gt: fiveMinutesAgo } },
              { $push: { messages: { from: fromUserId, type: type, message: message }  } }, 
              { multi: 1 },
               cb );
};


function sendToTabsOfUser( toUserId, excludeTabId, fromUserId,  type, message, cb ){
  var Tab = mongoose.model('Tab');

  var fiveMinutesAgo = new Date();
  fiveMinutesAgo.setMinutes( fiveMinutesAgo.getMinutes() - 5 );

  // The callback needs to be there full stop
  cb = cb || function(){} 

  Tab.update( { userId: toUserId, _id: { $ne: excludeTabId}, lastPing: { $gt: fiveMinutesAgo } },
              { $push: { messages: { from: fromUserId, type: type, message: message }  } }, 
              { multi: 1 },
               cb );
}

function clearMessagesOfTab( tabId, cb ){
  var Tab = mongoose.model('Tab');

  // The callback needs to be there full stop
  cb = cb || function(){} 

  Tab.update( { _id: tabId }, { messages: [] }, cb );
}


function registerPing( req, res, next ){
  var sendResponse = hotplate.getModule('hotProtocol').sendResponse;
  var e = hotplate.getModule('hotError').errors;

  registerTab( req.application.userId, req.application.workspaceId, function( err, doc ){
    if( err ){
      next(new e.RuntimeError( err ) );
    } else {
      sendResponse( res, { tabId: doc._id } );
    }
  });
}

function pollPing( req, res, next ){

  var sendResponse = hotplate.getModule('hotProtocol').sendResponse;
  var e = hotplate.getModule('hotError').errors;
  var tabId = req.params.tabId;
  var Tab = mongoose.model('Tab');

  // Get the messages from Json, safely
  try {
    if( ! req.body.messages )
      throw( new Error("req.body.messages not there") );
    var messages = JSON.parse(req.body.messages);
  } catch(e) {
    var messages = [];
  }

  // Usual sanity checks
  if( ! hotplate.mongoIdCheck(tabId) ){
    next( new e.ValidationError( "Tab id not valid" ) );
    return;
  }

  tabId = mongoose.Types.ObjectId(tabId);
  placeMessagesInQueue( function(err ){
    if( err ){
      next(new e.RuntimeError( err ) );
    } else {
      returnMessages();
    };
  })


  function placeMessagesInQueue( cb ){

    // Send data out first
    if( messages && Array.isArray(messages) ){

     messages.forEach( function(item) {
        if( item.type !== 'reset' ){
          if( item.toType === 'toMyTabs'){
            sendToTabsOfUser( req.application.userId, tabId,      req.application.userId, item.type, item.message );
          }
          if( item.toType === 'toEverybody'){
            sendToTabsOfWorkspace( req.application.workspaceId, tabId,      req.application.userId, item.type, item.message );
          }
        }
      });

    } 

    cb(null);
  }


  function returnMessages(){

    // Attempt to retrieve the tab's info
    retrieveTab( tabId, function( err, doc ){
      if( err ){
        next(new e.RuntimeError( err ) );
      } else {

        // Document returned, check that it's owned by the caller, send messages and clear
        if( doc ){
          sendResponse( res, doc.messages );
          clearMessagesOfTab( tabId ); // Nothing we can do if this doesn't work

        // Nothing returned: tab doesn't exist, creating a new one and letting client know
        } else {
          registerTab(  req.application.userId, req.application.workspaceId, function( err, doc ){
            if( err ){
              next(new e.RuntimeError( err ) ); 
            } else {
              var messages = [];
              messages.push( { from: req.application.userId, type: 'reset', message: { tabId: doc._id } } );
              sendResponse( res, messages );
            }
          });
        }
      }
    });
  }

}



        //  if( ! doc.workspaceId.equals( req.application.workspaceId )  ){
        //    next( new e.ForbiddenError() );
        //  } else {
        //  ...
        //  }
