// hotError/lib/hotError.js
/*!
 * Module dependencies.
 */

var util = require('util')
  , hotplate = require('hotplate')
  , path = require('path')
  , logger = require('hotServerLogger')
  , e = require('allhttperrors')
;



// Sets logLEvels for errors, _and_ sets some more friendly defaults in terms of messages etc.

Object.keys( e ).forEach( function( o ){

  switch( o.name ){
    case 'ServiceUnavailableError':
      o.logLevel = 5;
      o.name = "Server error";
    break;

    case 'BadRequestError':
      o.logLevel = 2;
    break;

    case 'UnprocessableEntityError':
      o.logLevel = 2;
      o.name = "Validation error";
    break;

  }

})


exports.hotErrorHandler = function( err, req, res, next){

  var sendResponse = hotplate.getModule('hotProtocol').sendResponse;
  var accept = req.headers.accept || '';
  var env = process.env.NODE_ENV || 'development';


  // Print the stack to console in any case
  if( env !== 'test' ) console.error( err.stack );

  // It's not an HTTP error: make a "system" log entry for it, as this
  // should never ever happen. This entry will be on TOP of the "normal" entry
  // (which the user will be able to see)
  if( typeof( e[ err.name ] ) === 'undefined' ){
    // Log the error
    logger.log({
      logLevel     : 5,
      errorName    : err.name,
      message      : err.message,
      errors       : '',
      system       : 1,
      data         : { error: err, trace: err.stack },
    });
  }

  // It's not an HTTP error: make up a new one, and incapsulate original error in it
  if( typeof( e[ err.name ] ) === 'undefined'  ){
    err = new self.ServiceUnavailableError( { originalErr: err } );
  }

  // Log the error, in 
  logger.log({
    logLevel     : err.logLevel,
    errorName    : err.name,
    message      : err.message,
    errors       : err.errors,
    system       : 0,
    data         : { error: err },
  });


  // If it was an HTML page that generated the error, return hotplate's error page
  if( ~accept.indexOf('html') ){
    req.application.error = err;
    hotplate.get('errorPage')(req, res, next);    

  // If it was a JSON request, do a sendResponse following the protocol
  } else if (~accept.indexOf('json')) {
    sendResponse( res, { message: err.message, errors: err.errors }, err.httpError );

  // Otherwise, just print the output out...
  } else {
    res.writeHead(err.httpError, { 'Content-Type': 'text/plain' });
    res.end( err.message );
  }  
  
}
