// bd/lib/bd.js
var 
  dummy
, hotplate =  require('hotplate')
, path = require('path')
, mongoose = require('mongoose')
;

app = hotplate.app;

exports.hotHooks = hooks = {}

hooks.run = function( done ){

  app.get(  '/call/workspacesAll', getWorkspacesAll );
  done( null );
}


hooks.pageElementsPerPage = function( done, elements, req, pageName ){

  if( pageName  === 'hotDojoAppContainer/container' ){

    done( null, {
      moduleName: 'bd', result: {
        jses: ['bd.js'],
        csses: ['bd.css']
      }
    });
  } else {
    done( null, { moduleName: 'bd', result: {} }  );
  }

}

// Simply activate path to client files
hooks.clientPaths = function( done ){
  done( null, { moduleName: 'bd', result: [ path.join(__dirname, '../client') ] } );
}

hooks.stores = function( done ){
  done( null, {
    workspacesAll: { target: '/call/workspacesAll', idProperty: '_id', sortParam: 'sortBy', },
  });
}

function getWorkspacesAll( req, res, next ){
 
  var name = req.query.name;
  var sendResponse = hotplate.getModule('hotProtocol').sendResponse;
  var e = hotplate.getModule('hotError').errors;

  // Looks for a workspace. If it's there, answers without errors. If not,
  // answers with a very short error
  Workspace = mongoose.model('Workspace');
  Workspace.find({}, function(err, doc){
    if(err ){
      next(new e.RuntimeError( err ) );
    } else {

      sendResponse( res, { data: doc  } );
    }
  });
}
