// mongoUtil/index.js

var 
  dummy
, hotplate =  require('hotplate')
, mongoose = require('mongoose')
;

app = hotplate.app;

exports.hotHooks = hooks = {}


exports.trim = function(length){
  return function(v){
    console.log("Called trim for: %s", v);
    return v.substring(0,length);
  }
}

exports.lowerCase = function(){
  return function(v){
    return v.toLowerCase();
  }
}

exports.upperCase = function(){
  return function(v){
    return v.toUpperCase();
  }
}

exports.trimAndLowerCase = function(length){
  return function(v){
    console.log("CALLED ON: %s", v);
    return v.substring(0, length).toLowerCase();
  }
}
exports.trimAndUpperCase = function(length){
  return function(v){
    return v.substring(0, length).toUpperCase();
  }
}

/* UNUSED 
e.g.
 mongoUtil.addSharedValidator( 'Access', 'settings.localEmail', 'hotCommonValidators', 'email' );

exports.addSharedValidator = function(modelName, path, module, sharedFunc ){
  var functions = hotplate.getModule('hotSharedCode').sharedFunctions;

  mongoose.model(modelName).schema.path(path).validate(
    function(v) {  return functions[module][sharedFunc](v); },
    functions[module][sharedFunc](false)  
  );
}
*/
