// hotClientFiles/lib/hotClientFiles.js

/* TODO:
*/

/*!
 * Module dependencies.
 */

var util = require('util')
, fs = require('fs')
, path = require('path')
, send = require('send')
, hotplate = require('hotplate')
;


/**
 * HotClientFiles constructor.
 *
 * The exports object of the `HotClientFiles` module is an instance of this class.
 * Most apps will only use this one instance.
 *
 * @api public
 */

function HotClientFiles() {
  this.localLocations = {};
};


/**
 * The exports object is an instance of HotClientFiles.
 *
 * @api public
 */

module.exports = exports = new HotClientFiles;
var hotClientFiles = module.exports;


/**
 * The HotClientFiles constructor
 *
 * The exports of the mongoose module is an instance of this class.
 *
 * ####Example:
 *
 *     var hotClientFiles= require('hotplate');
 *     var hotClientFiles2 = new hotpage.HotClientFiles();
 *
 * @api public
 */

hotClientFiles.HotClientFiles = HotClientFiles;

HotClientFiles.prototype.hotHooks = {};


HotClientFiles.prototype.hotHooks.init = function( done ){

  var that = this;
  var moduleName;

  hotplate.invokeAll('clientPaths', function(err, results){
    results.forEach( function(element) {
      moduleName = element.moduleName;

      element.result.forEach( function( clientPath ){
        that.localLocations[ moduleName ] = clientPath;
      });
    });

    done( null );
  });

}
HotClientFiles.prototype.hotHooks.init.invokes = [ 'clientPaths' ];


HotClientFiles.prototype.add = function(moduleName, localLocation){
  this.localLocations[ moduleName ] = localLocation
}

HotClientFiles.prototype.serve = function(options){
  var that = this;

  options = options || {};

  // Make up the regular expression to check if the URL requested in
  // one of the URLs monitored by the module
  var staticUrlPathRegExp = new RegExp('^' + hotplate.get('staticUrlPath') + '/(.*?)/(.*)');

  // Return the actual route
  return function static(req, res, next) {

    // If there is a match between the request and the
    // expected sub-url...
    var  match = req.path.match( staticUrlPathRegExp );

    // hotplate.log("Match is: " , match);
    // hotplate.log("localLocations is: " + that.localLocations[ match[1] ] );

    if( match && that.localLocations[ match[1] ] ){

      // Create the variable names
      var moduleName = match[1];
      var fileLocation = match[2];

      // Get the "local" variable for that module
      var localDir = that.localLocations[ moduleName ];

      // At this point, if:
      //  - req.path      is /hotPlateDir/testingModule/css/one.css
      //  - staticUrlPath is /hotPlateDir 
      //  
      // Then:
      //  - localDir     would be  /home/www/program/node_modules/aModule/public_files/
      //  - fileLocation would be  css/one.css 

      // Send it!
      function error(err) {
        if( 404 == err.status) return next();
        next(err);
      }
      send(req, fileLocation )
       .maxage(options.maxAge || 0)
       .root(localDir)
       .hidden(options.hidden)
       .on('error', error )
       .pipe(res);
    } else {
      next();
    }
    
  }

};
