// hotDojoAppContainer/lib/hotDojoAppContainer.js

var dummy
  , path = require( 'path' )
  , hotplate = require( 'hotplate' )
;

var hooks = exports.hotHooks = {} ;


hooks.init = function( done ){

  // Global variables (global at module level)
  var dummy
  , logger = hotplate.getModule( 'hotServerLogger' )
  // , hotPage = hotplate.getModule( 'hotPage' )
  , sendResponse = hotplate.getModule('hotProtocol').sendResponse
;

  hotplate.set('errorPage', function( req, res, next ){
    res.send("ERROR!")
  });

  done( null );
}


hooks.run = function( done ){
  var sendResponse = hotplate.getModule('hotProtocol').sendResponse;

  hotplate.app.get( '/ws',                      function(req, res, next){ res.redirect('/pages/login'); } );
  hotplate.app.get( '/ws/:workspaceIdPages',    mainApp);

  // Example routes defined, containing :tokenCall and :workspaceIdCall
  // app.get(      '/api/1/:tokenCall/users', testApi1 );
  // app.get( '/call/:workspaceIdCall/users', testApi1 );

  hotplate.app.get('/api/1/:workspaceIdCall/users', function(req, res, next ){
    sendResponse( res, { data: [ { result: 'OK!' } ]  } );
  });


  done( null );
} 

hooks.clientPaths = function( done ){
  done( null, { moduleName: 'hotDojoAppContainer', result:[ path.join( __dirname, '../client' ) ] } );
}

hooks.stores = function( done ){
  done( null, {
    users: { target: '/api/1/:workspaceIdCall/users',      idProperty: '_id', sortParam: 'sortBy', },
  });
}


function mainApp( req, res ){

  var hotPage = hotplate.getModule('hotPage');
  var sendResponse = hotplate.getModule('hotProtocol').sendResponse;

  var extraJses = (new hotPage.Jses).add( 'hotDojoAppContainer', 'main.js' );
  var extraCsses = (new hotPage.Csses).add( 'hotDojoAppContainer', 'main.css' );
  var extraVars = new hotPage.Vars;
  extraVars.add( 'hotDojoAppContainer', 'login', req.session.login );
  extraVars.add( 'hotDojoAppContainer', 'workspaceId', req.application.workspaceId );
  extraVars.add( 'hotDojoAppContainer', 'workspaceName', req.application.workspaceName );

  hotPage.processPageTemplate(
    {
      jses: extraJses,
      vars: extraVars,
      // csses:extraCsses,
      body: '<body class="claro"><div id="appMainScreen"></div></body>',
    },
    req, 
    'hotDojoAppContainer/main',
    function( result ){ 
      res.send( result );
      logger.log( { message: "App main screen page served" } );
    }
  );
  return;
}

hooks.pageElementsPerPage = function( done, elements ){

  console.log("CALLED pageElementsPerPage for appContainer");

  done( null, { moduleName: 'hotDojoAppContainer', result: {
    vars:  [ { name: 'ATTEMPT', value: 'the value' } ],
    csses:   [  'main.css'  ],
  } } );
 

}
