// hotPage/lib/hotPage.js

/* TODO:
*/

/*!
 * Module dependencies.
 */

var dummy
, util = require('util')
, fs = require('fs')
, path = require('path')
, send = require('send')

, hotplate = require('hotplate')
, Vars = require('./Vars.js')
, Csses = require('./Csses.js')
, Jses = require('./Jses.js')
, HeadLines = require('./HeadLines.js')
;


/**
 * HotPage constructor.
 *
 * The exports object of the `HotPage` module is an instance of this class.
 * Most apps will only use this one instance.
 *
 * @api public
 */

function HotPage() {

  // Make these objects available to other modules too
  // with hotPage.Vars, hotPage.Csses, hotPage.Jses
  this.Vars = Vars;
  this.Csses = Csses;
  this.Jses = Jses;
  this.HeadLines = HeadLines;

  this.vars = new Vars;
  this.csses = new Csses;
  this.jses = new Jses;
  this.headLines = new HeadLines;

  // The page template
  this.pageTemplate = "<!DOCTYPE HTML>\n<html>\n<head><meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\" /><title>[[TITLE]]</title>\n[[HEAD]]\n</head>\n[[BODY]]\n</html>\n";


};


/**
 * The exports object is an instance of HotPage.
 *
 * @api public
 */

module.exports = exports = new HotPage;
var hotPage = module.exports;


/**
 * The HotPage constructor
 *
 * The exports of the mongoose module is an instance of this class.
 *
 * ####Example:
 *
 *     var hotPage= require('hotplate');
 *     var hotPage2 = new hotpage.HotPage();
 *
 * @api public
 */

hotPage.HotPage = HotPage;


HotPage.prototype.processPageTemplate = function( elements, leavePlaceholders ) {

  elements = elements || {};

  var r = this.pageTemplate;

  // Replace the elements: csses, jses and vars will go where [[HEAD]] is,
  // the title will go wheer [[TITLE]] is, the body where [[BODY]] is
  if ( elements.headLines ) r = r.replace(/(\[\[HEAD\]\])/,  elements.headLines.render() + '$1' );
  if ( elements.vars)       r = r.replace(/(\[\[HEAD\]\])/,  elements.vars.render() + '$1' );
  if ( elements.csses )     r = r.replace(/(\[\[HEAD\]\])/,  elements.csses.render() + '$1' );
  if ( elements.jses )      r = r.replace(/(\[\[HEAD\]\])/,  elements.jses.render() + '$1' );

  if ( elements.title ) r = r.replace(/(\[\[TITLE\]\])/, elements.title + '$1' ); 
  if ( elements.body )  r = r.replace(/(\[\[BODY\]\])/,  elements.body + '$1' ); 
  
  // Take placeholders away. The template is probably being processed by a page,
  // which most likely added its own title, csses, js, etc.
  if( ! leavePlaceholders){
      r = r.replace(/\[\[(HEAD|TITLE|BODY)\]\]/g, '');
  }

  return r;
}

HotPage.prototype.hotHooks = {};


HotPage.prototype.hotHooks.init = function( done ){

  var that = this;
  var moduleName;


  hotplate.invokeAll('pageElements', function(err, results){

    console.log("Results: ");
    console.log(results);

    results.forEach( function(element) {
      moduleName = element.moduleName;
      result = element.result;

      for( var type in result){
        var elements = result[type];
        switch( type ) {

          case 'jses':
           elements.forEach( function(element) { 
              that.jses.add(moduleName, element);
            });
          break;

          case 'headLines':
            elements.forEach( function(element) { 
              that.headLines.add(moduleName, element );
            });
          break;

          case 'csses':
            elements.forEach( function(element) { 
              that.csses.add(moduleName, element);
            });
          break;

          case 'vars':
            elements.forEach( function(element) { 
              that.vars.add(moduleName, element.name, element.value );
            });
          break;
        }
      }
    })
  });
  
  // Renders the template with the values returned by the hooks...
  this.pageTemplate = this.processPageTemplate( { csses:this.csses, jses:this.jses, vars:this.vars, headLines:this.headLines } , true );
  done( null );

}

HotPage.prototype.hotHooks.init.invokes = [ 'pageElements' ];
