// hotCoreStoreLogger/lib/hotCoreStoreLogger.js

var util = require('util')

  , hotplate = require('hotplate')
  , ObjectId = require('mongowrapper').ObjectId
  , declare = require('simpledeclare')
  , Store = require('jsonreststores')
  , MongoDriverMixin = require('jsonreststores/MongoDriverMixin.js')
  , app = hotplate.app
  , db = hotplate.get('db')

  , Schema = require('simpleschema')
  , MongoSchemaMixin = require('simpleschema/MongoSchemaMixin.js')
;


// Sets the right mixins for Store and Schema
Schema = declare( [ Schema, MongoSchemaMixin ] );
Store = declare( [ Store, MongoDriverMixin ] );

// Module's variables
// var logFields = {};
var LogStore = null;

var hooks = exports.hotHooks = {}

hooks.initLoggers = function( logFields, done ){

  var s = new Schema( logFields );

  // MAYBE: Manipulate schema here if needed: add things so that they are marked as sortable etc.

  LogStore = declare( Store,  {
    storeName: 'logs',
    db: db,
    schema: s,
    paramIds: [ 'logId' ],
  });

  done( null ); 
}

hooks.log = function( entry, done ){

  console.log("***************************** hotStoreLogger is logging:");
  console.log( entry );

  // Add the entry. This might fail. If it does, it will do so quietly
  LogStore.Post( entry );
  done( null );
}

hooks.stores = function( done ){

  done( null,  {
    moduleName: 'hotCoreStoreLogger', result: [ LogStore ]
  });

}
